server/mqtt.go

```diff
@@ -3144,7 +3154,52 @@ func mqttPubTrace(pp *mqttPublish) string {
 //
 // Runs from the client's readLoop.
 // No lock held on entry.
-func (s *Server) mqttProcessPub(c *client, pp *mqttPublish) error {
+func (s *Server) mqttProcessPub(c *client, pp *mqttPublish, trace bool) error {
+	qos := mqttGetQoS(pp.flags)
+
+	// Forward the PUBLISH packet on, unless it is a QoS2 message that is not complete yet.
+	drop := false
+	if qos == 2 {
+		c.mqtt.sess.mu.Lock()
+		if len(c.mqtt.sess.pendingQOS2Received) > 0 {
+			_, drop = c.mqtt.sess.pendingQOS2Received[pp.pi]
+		}
+		c.mqtt.sess.mu.Unlock()
+	}
+
+	if !drop {
+		if err := s.mqttForwardPub(c, pp); err != nil {
```

Comment:
If I read this right, for QoS2, you seem to store message (for the first PUBLISH received), but that will also initiate delivery, then send the PUBREC then "store" the PI.  I write "store" because this is just in memory and not really persisted anywhere. 
The spec seem to indicate that there are methods A or B, and this looks to me a mix of the two. (see figure 4.3 [here](https://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718102)).

Since we don't really "store" the PI, then if the server were to crash after "mqttForwardPub" (which stores and initiates delivery), but before sending the PUBREC, then if the sender were to call again PUBLISH with same PI, we would violate I believe this rule: `Until it has received the corresponding PUBREL packet, the Receiver MUST acknowledge any subsequent PUBLISH packet with the same Packet Identifier by sending a PUBREC. It MUST NOT cause duplicate messages to be delivered to any onward recipients in this case`.

Reply:
Right. For PUBACKs, Is this accomplished (in reverse) with `cpending`? I have not understood that part of the code yet, nor the session persistence/management code yet.

Reply:
Re-did using subject-based deduplication. Resolving the comment as no longer applicable.